framework/src/meshgenerators/PatternedMeshGenerator.C

```diff
@@ -88,35 +116,74 @@ PatternedMeshGenerator::generate()
-                                              getParam<BoundaryName>("top_boundary"),
-                                              getParam<BoundaryName>("bottom_boundary")};
+    // Keep track of used IDs so we can later find IDs that are unused across all meshes
+    used_boundary_ids.insert(ids.begin(), ids.end());
 
-  std::vector<boundary_id_type> ids =
-      MooseMeshUtils::getBoundaryIDs(*_meshes[0], boundary_names, true);
+    // Keep track of unique boundary ids seen in each input mesh
+    std::set<boundary_id_type> id_set;
+    // Keep track of indices of first instance of each unique boundary id
+    std::map<boundary_id_type, size_t> seen_bid_to_index_map;
+
+    size_t index = 0;
+    for (const auto side : make_range(4))
+    {
+      // Check if the boundary has been initialized
+      if (ids[side] == Moose::INVALID_BOUNDARY_ID)
+        paramError("inputs",
+                   "The '",
+                   boundary_param_names[side],
+                   "' parameter with value '",
+                   boundary_names[side],
+                   "' does not exist in input mesh '",
+                   _input_names[i],
+                   "'");
+
+      input_bids[i][side] = ids[side];
+
+      // We only do this when i == 0 because all input meshes should have the
+      // same index map. Allowing different index maps for different input
+      // meshes results in undefined behaviour when stitching
+      if (i == 0)
+      {
+        if (!id_set.count(ids[side]))
+        {
+          input_bids_unique[i].push_back(ids[side]);
+          id_set.insert(ids[side]);
+          seen_bid_to_index_map[ids[side]] = index;
+          boundary_name_2_index_map[boundary_param_names[side]] = index++;
+        }
+        else
+          boundary_name_2_index_map[boundary_param_names[side]] = seen_bid_to_index_map[ids[side]];
+      }
 
-  mooseAssert(ids.size() == boundary_names.size(),
-              "Unexpected number of ids returned for MooseMeshUtils::getBoundaryIDs");
+      else // i > 0
+      {
+        if (ids[side] != input_bids[i - 1][side])
+          have_common_ids = false;
+
+        if (!id_set.count(ids[side]))
+        {
+          input_bids_unique[i].push_back(ids[side]);
+          id_set.insert(ids[side]);
+        }
+      }
+    }
 
-  boundary_id_type left = ids[0], right = ids[1], top = ids[2], bottom = ids[3];
+    lengths_set.insert(input_bids_unique.size());
+  }
 
-  // Check if all the boundaries have been initialized
-  if (left == -123)
-    mooseError("The left boundary has not been initialized properly.");
-  if (right == -123)
-    mooseError("The right boundary has not been initialized properly.");
-  if (top == -123)
-    mooseError("The top boundary has not been initialized properly.");
-  if (bottom == -123)
-    mooseError("The bottom boundary has not been initialized properly.");
+  // Error check on lengths of input_bids_unique
+  if (lengths_set.size() != 1)
```

Comment:
you could check the content to be 1 or 4 just below

Reply:
See our above conversation. I am leaving this check out for more user flexibility.